cmd/cloudFoundryDeleteService.go

```diff
@@ -1,48 +1,108 @@
 package cmd
 
 import (
+	"bytes"
+	"fmt"
+	"strings"
+
 	"github.com/SAP/jenkins-library/pkg/command"
 	"github.com/SAP/jenkins-library/pkg/log"
 	"github.com/SAP/jenkins-library/pkg/telemetry"
 )
 
-func cloudFoundryDeleteService(CloudFoundryDeleteServiceOptions cloudFoundryDeleteServiceOptions, telemetryData *telemetry.CustomData) error {
+func cloudFoundryDeleteService(options cloudFoundryDeleteServiceOptions, telemetryData *telemetry.CustomData) error {
 
 	c := command.Command{}
 
 	// reroute command output to logging framework
 	c.Stdout(log.Entry().Writer())
 	c.Stderr(log.Entry().Writer())
 
-	cloudFoundryLogin(CloudFoundryDeleteServiceOptions, &c)
+	var err error
 
-	err := cloudFoundryDeleteServiceFunction(CloudFoundryDeleteServiceOptions.CfServiceInstance, &c)
-	if err != nil {
+	err = cloudFoundryLogin(options, &c)
+
+	if err == nil && options.CfDeleteServiceKeys == true {
+		err = cloudFoundryDeleteServiceKeys(options, &c)
+	}
+
+	if err == nil {
+		err = cloudFoundryDeleteServiceFunction(options.CfServiceInstance, &c)
+	}
+
+	if err == nil {
+		err = cloudFoundryLogout(&c)
+		if err != nil {
+			log.Entry().
+				WithError(err).
+				Fatal("Error while logging out occured.")
+		}
+	} else if err != nil {
 		cloudFoundryLogout(&c)
 		log.Entry().
 			WithError(err).
-			Fatal("Failed to delete Service!")
-		return err
+			Fatal("Error occured.")
 	}
 
-	cloudFoundryLogout(&c)
+	return err
+}
+
+func cloudFoundryDeleteServiceKeys(options cloudFoundryDeleteServiceOptions, c execRunner) error {
+
+	log.Entry().Info("Deleting inherent Service Keys")
 
-	return nil
+	var cfFindServiceKeysScript = []string{"service-keys", options.CfServiceInstance}
+
+	var serviceKeyBytes bytes.Buffer
+	c.Stdout(&serviceKeyBytes)
+
+	err := c.RunExecutable("cf", cfFindServiceKeysScript...)
+	if err != nil {
+		return fmt.Errorf("Failed to Delete Service Key, most likely your service doesn't exist: %w", err)
+	}
+
+	if len(serviceKeyBytes.String()) > 0 {
+		var lines []string = strings.Split(serviceKeyBytes.String(), "\n")
+		if len(lines) <= 4 {
+			log.Entry().Info("No Service Keys active to be deleted")
+			return err
+		}
+		var numberOfLines = len(lines)
+		log.Entry().WithField("Number of service keys :", numberOfLines-4).Info("ServiceKey")
+
+		//Deleting all matched Service Keys for Service
+		for i := 3; i <= numberOfLines-2; i++ {
+			log.Entry().WithField("Deleting Service Key", lines[i]).Info("ServiceKeyDeletion")
+
+			var cfDeleteServiceKeyScript = []string{"delete-service-key", options.CfServiceInstance, lines[i], "-f"}
+
+			err := c.RunExecutable("cf", cfDeleteServiceKeyScript...)
+			if err != nil {
+				return fmt.Errorf("Failed to Delete Service Key: %w", err)
+			} else {
+				log.Entry().Info("ServiceKeys have been deleted!")
+			}
+		}
+		return err
+	} else {
+		log.Entry().Info("No service key could be retrieved for your requested Service")
+		return err
+	}
```

Comment:
```suggestion
	if len(serviceKeyBytes.String()) == 0 {
		log.Entry().Info("No service key could be retrieved for your requested Service")
		return nil
	}

	var lines []string = strings.Split(serviceKeyBytes.String(), "\n")
	if len(lines) <= 4 {
		log.Entry().Info("No Service Keys active to be deleted")
		return err
	}
	var numberOfLines = len(lines)
	log.Entry().WithField("Number of service keys :", numberOfLines-4).Info("ServiceKey")

	// Deleting all matched Service Keys for Service
	for i := 3; i <= numberOfLines-2; i++ {
		log.Entry().WithField("Deleting Service Key", lines[i]).Info("ServiceKeyDeletion")

		var cfDeleteServiceKeyScript = []string{"delete-service-key", options.CfServiceInstance, lines[i], "-f"}

		err := c.RunExecutable("cf", cfDeleteServiceKeyScript...)
		if err != nil {
			return fmt.Errorf("Failed to Delete Service Key: %w", err)
		} else {
			log.Entry().Info("ServiceKeys have been deleted!")
		}
	}
	return err
```

Reply:
Sorry for the weird diff, I just tried to align the happy-path.

Reply:
Changed it accordingly for the happy-path